### PR TITLE
fix: biencoder bidirectional masking and consistent attn_implementation default

### PR DESCRIPTION
**Pull request by oliverholworthy**
- Replace `_update_causal_mask` with `create_bidirectional_mask` in `LlamaBidirectionalModel` to fix bidirectional attention with newer transformers versions and non-flash attention backends
- Add `attn_implementation` option to the mining recipe so the attention backend can be configured at inference time
- Default `attn_implementation` to None (sdpa) instead of flash_attention_2 in NeMoAutoModelBiencoder

## Motivation

The old `_update_causal_mask` override had three problems:

1. Dead method on transformers >= 4.53: upstream removed `_update_causal_mask` from LlamaModel, so our override was never called and the model silently fell back to causal masking with sdpa/eager backends.
2. Wrong mask format for SDPA/eager: it returned None or passed through the raw 2D mask without expanding to 4D float, causing shape/dtype errors in scaled_dot_product_attention.
3. No backend awareness: Flash Attention 2 expects a 2D (batch, seq_len) mask, while SDPA/eager need a 4D (batch, 1, seq_len, seq_len) float mask. The old code didn't distinguish between them.

The new _create_bidirectional_mask:
- Uses native `create_bidirectional_mask` from transformers.masking_utils when available (transformers >= 5.0)
- Falls back to backend-aware logic for older transformers: 2D passthrough for flash_attention_2, _prepare_4d_attention_mask for SDPA/eager

# Before your PR is "Ready for review"

**Pre checks**:

- [X] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA-NeMo/Automodel/blob/main/CONTRIBUTING.md)
- [X] Did you write any new necessary tests?
- [X] Did you add or update any necessary documentation?

If you haven't finished some of the above items you can still open "Draft" PR.

### CODE REVIEW

File: nemo_automodel/_transformers/auto_model.py [hunk 1]
```diff
@@ -734,7 +734,7 @@ def from_pretrained(
         share_encoder: bool = True,
         pooling: str = "avg",
         l2_normalize: bool = True,
-        attn_implementation: str = "flash_attention_2",
+        attn_implementation: Optional[str] = None,
```
Comment (adil-a) [inline nemo_automodel/_transformers/auto_model.py]: why shouldn't we use FA2 by default? we leave mem savings and speed on the table otherwise

File: nemo_automodel/_transformers/auto_model.py
Comment (adil-a) [inline nemo_automodel/_transformers/auto_model.py:861]: we have retry methods if fa2 doesn't work here btw

File: nemo_automodel/components/models/llama_bidirectional/model.py [hunk 2]
```diff
@@ -33,6 +33,16 @@
 from transformers.processing_utils import Unpack
 from transformers.utils import TransformersKwargs
 
+# Check if native create_bidirectional_mask exists (transformers >= 5.0)
+try:
```
Comment (adil-a) [inline nemo_automodel/components/models/llama_bidirectional/model.py]: can we enforce v5 now? automodel requires it so I think this code block is unnecessary

File: nemo_automodel/components/models/llama_bidirectional/model.py [hunk 3]
```diff
@@ -94,15 +104,45 @@ def __init__(self, config: LlamaConfig):
...
+                tokens and 0 for padding, or None.
+
+        Returns:
+            A 4D float mask for sdpa/eager, a 2D mask for flash_attention_2,
+            or None when no masking is needed.
+        """
+        if attention_mask is None:
+            return None
+
+        if _HAS_NATIVE_BIDIRECTIONAL_MASK:
```
Comment (adil-a) [inline nemo_automodel/components/models/llama_bidirectional/model.py]: same as above, let's just enforce v5 and use this

File: nemo_automodel/recipes/biencoder/mine_hard_negatives.py [hunk 4]
```diff
@@ -69,6 +69,8 @@
     # Model loading parameters (loaded directly, not from config)
     "model_name_or_path": None,  # Required: path to model checkpoint
     "tokenizer_name_or_path": None,  # Optional: defaults to model_name_or_path
+    # Attention implementation for model loading
+    "attn_implementation": None,  # None = use model default; "sdpa", "flash_attention_2", "eager"
```
Comment (adil-a) [inline nemo_automodel/recipes/biencoder/mine_hard_negatives.py:73]: same comment as above as auto_model.py

File: nemo_automodel/components/models/llama_bidirectional/model.py (hunk 2 above)
Comment (oliverholworthy) [reply nemo_automodel/components/models/llama_bidirectional/model.py]: this code I believe ends up in the checkpoint we save to disk, so while the automodel library my dependon on >=5, an open question may be whether the checkpoint we save we want to support earlier versions

File: nemo_automodel/_transformers/auto_model.py (hunk 1 above)
Comment (oliverholworthy) [reply nemo_automodel/_transformers/auto_model.py]: For consistency i've changed to `DEFAULT_ATTN_IMPLEMENTATION` to match the other classes in this file, which uses `flash_attention_2` if the flash_attn package is installed (falling back to sdpa otherwise).

File: nemo_automodel/components/models/llama_bidirectional/model.py (hunk 2 above)
Comment (oliverholworthy) [reply nemo_automodel/components/models/llama_bidirectional/model.py]: We can revisit the checkpoint portability question in a follow-up, For now I've changed to a direct import as suggested.

File: nemo_automodel/components/models/llama_bidirectional/model.py (hunk 3 above)
Comment (oliverholworthy) [reply nemo_automodel/components/models/llama_bidirectional/model.py]: updated

File: nemo_automodel/_transformers/auto_model.py
Comment (oliverholworthy) [reply nemo_automodel/_transformers/auto_model.py:861]: In what scenario would this Exception handler be reached?

File: nemo_automodel/recipes/biencoder/mine_hard_negatives.py (hunk 4 above)
Comment (oliverholworthy) [reply nemo_automodel/recipes/biencoder/mine_hard_negatives.py:73]: This is intentionally None, since the Model implementation determines it's own default. We only pass a value through to the model class if this is not None
